gdal/port/cpl_vsil_plugin.cpp

```diff
@@ -0,0 +1,356 @@
+{
+    delete m_cb;
+}
+
+
+
+VSIVirtualHandle* VSIPluginFilesystemHandler::Open( const char *pszFilename,
+                                                  const char *pszAccess,
+                                                  bool bSetError )
+{
+    if( !IsValidFilename(pszFilename) )
+        return nullptr;
+    void *cbData = m_cb->open(m_cb->pUserData, GetCallbackFilename(pszFilename), pszAccess);
+    if (cbData == nullptr) {
+        if (bSetError) {
+            VSIError(VSIE_FileError, "%s: %s", pszFilename, strerror(errno));
+        }
+        return nullptr;
+    }
+    return new VSIPluginHandle(this, cbData);
+}
+
+const char* VSIPluginFilesystemHandler::GetCallbackFilename(const char *pszFilename) {
+    return pszFilename + strlen(m_Prefix);
+}
+
+bool VSIPluginFilesystemHandler::IsValidFilename(const char *pszFilename) {
+    if( !STARTS_WITH_CI(pszFilename, m_Prefix) )
+        return false;
+    return true;
+}
+
+
+int VSIPluginFilesystemHandler::Stat( const char *pszFilename,
+                                      VSIStatBufL *pStatBuf,
+                                      int nFlags )
+{
+    if( !IsValidFilename(pszFilename) ) {
+        errno = EBADF;
+        return -1;
+    }
+
+
+    memset(pStatBuf, 0, sizeof(VSIStatBufL));
+
+    int nRet = 0;
+    if ( m_cb->stat != nullptr ) {
+        nRet = m_cb->stat(m_cb->pUserData, GetCallbackFilename(pszFilename), pStatBuf, nFlags);
+    } else {
+        nRet = -1;
+    }
+    return nRet;
+}
+
+int VSIPluginFilesystemHandler::Seek(void *pFile, vsi_l_offset nOffset, int nWhence) {
+    if (m_cb->seek != nullptr) {
+        return m_cb->seek(pFile,nOffset,nWhence);
+    }
+    CPLError(CE_Failure, CPLE_AppDefined, "Seek not implemented for %s plugin", m_Prefix);
+    return -1;
+}
+
+vsi_l_offset VSIPluginFilesystemHandler::Tell(void *pFile) {
+    if (m_cb->tell != nullptr) {
+        return m_cb->tell(pFile);
+    }
+    CPLError(CE_Failure, CPLE_AppDefined, "Tell not implemented for %s plugin", m_Prefix);
+    return -1;
+}
+
+size_t VSIPluginFilesystemHandler::Read(void *pFile, void *pBuffer, size_t nSize, size_t nCount) {
+    if (m_cb->read != nullptr) {
+        return m_cb->read(pFile, pBuffer, nSize, nCount);
+    }
+    CPLError(CE_Failure, CPLE_AppDefined, "Read not implemented for %s plugin", m_Prefix);
+    return -1;
+}
+
+int VSIPluginFilesystemHandler::HasOptimizedReadMultiRange(const char* /*pszPath*/ ) {
+    if (m_cb->read_multi_range != nullptr) {
+        return TRUE;
+    }
+    return FALSE;
+}
+
+VSIRangeStatus VSIPluginFilesystemHandler::GetRangeStatus( void *pFile, vsi_l_offset nOffset, vsi_l_offset nLength) {
+    if (m_cb->get_range_status != nullptr) {
+        return m_cb->get_range_status(pFile,nOffset,nLength);
+    }
+    return VSI_RANGE_STATUS_UNKNOWN;
+}
+
+int VSIPluginFilesystemHandler::ReadMultiRange( void *pFile, int nRanges, void ** ppData, const vsi_l_offset* panOffsets, const size_t* panSizes ) {
```

Comment:
I see that the behavior of joining consecutive ranges can be disabled with a config option in the vsicurl implementation, but fail to see when that could be desirable. Maybe this part could be moved up a bit higher in the vsi call tree?